ada/language/ast.py

```diff
@@ -3281,7 +3281,20 @@ def previous_part(go_to_incomplete=(Bool, True)):
                     pp.cast(T.BaseTypeDecl)._.is_private),
                 And(go_to_incomplete,
                     pp.is_a(T.IncompleteTypeDecl)),
-            )).cast(T.BaseTypeDecl))
+            )))._or(
+                # the previous part of this type is not in its semantic
+                # parents: this must mean we are in a package body and
+                # therefore the previous part must be in the private part of
+                # the package spec (RM 3.10.1).
```

Comment:
Wowow, quoting the RM and all :grin:. Two things:

1. The whole code you added seems unnecessary. This should be already handled by the above code (and it seems to be, since if I remove this chunk your test still passes).
2.  FIY, those types are commonly refered as "taft-amendment types". This would be good to add that to the comment, that will ring a bell more easily than a reference to the RM (which is a good thing to have anyway !)

Reply:
Good catch, I removed this part.